### ISSUE TITLE
Discussion: handling file extraction

### ISSUE THREAD

**Issue by erri120** [Nexus-Mods/NexusMods.App#245]
[`SevenZipExtractor`](https://github.com/Nexus-Mods/NexusMods.App/blob/main/src/ArchiveManagement/NexusMods.FileExtractor/Extractors/SevenZipExtractor.cs) is the only `IExtractor` we currently have, and it uses a native [7z binary](https://github.com/Nexus-Mods/NexusMods.App/tree/main/src/ArchiveManagement/NexusMods.FileExtractor/runtimes) to extract files. This approach has the following advantages and disadvantages:

**Advantages**:
- supports all formats we _currently_ need
- has cross-platform support
- unified way of extracting any archive using the CLI
- all the benefits of being an established tool: correctness, lots of support and knowledge available

**Disadvantages**:
- we need to start a new 7-Zip process for every archive we want to extract
- tracking and managing processes can be a pain
- we can't extract archives purely in-memory and can't make use of the [`IFileSystem`](https://github.com/Nexus-Mods/NexusMods.App/blob/main/docs/decisions/backend/0004-file-system-abstraction.md) abstraction
- unable to easily expand functionality if needed (like using different compression algorithms #241)

While our current situation works fine, only having the 7-Zip CLI as an option to extract archives might come and bite us later down the line. Being unable to extract archives in-memory is especially problematic, as it completely prevents `IFileSystem` from being used in the important places (file extraction and the data store). I believe now is a good time to look into this issue and create a [decision record](https://github.com/Nexus-Mods/NexusMods.App/blob/main/docs/decisions/project/0000-use-decision-records.md) based on the outcome.

As I see it, we have the following options (they are not mutually exclusive, since we can have multiple extractors):

1) continue using the 7-Zip CLI
2) use 3rd party libraries
3) develop our own libraries

All of these options come with their own set of benefits and problems, but the biggest decision driver will likely be the time commitment. Developing our own libraries would obviously fit our use case the most, but the time it takes to write, test, benchmark and optimize a library that has feature parity with 7-Zip and is not too much slower, will likely be immense.

Conversely, trying to find good 3rd party libraries and incorporating them into our code base has a similar time commitment in addition to forcing us to use very specific APIs.

Overall, file extraction is at the core of the app, and we resolve this issue in our current early staged of development.

**Comment by Aragas** [comment]
[sharpcompress](https://github.com/adamhathcock/sharpcompress) could used as a fully managed alternative.
Works good enough in our case, we're able to read every archive hosted on NexusMods for Bannerlord

**Comment by Sewer56** [comment]
> Yes, shelling out to 7zip is more of a "last resort" than the preferred method. The basic problem is that there are many different format types that are handled or not handled in various situations based on the combination of archives and extractors. For example, until last month several Skyrim mods could not be extracted with `sharpcompress` due to this bug ([adamhathcock/sharpcompress#526](https://github.com/adamhathcock/sharpcompress/issues/526)) Thankfully that was recently fixed so we can look at using the library again.

Aaaaaaa finally. I was writing a full big comment about this; and I remember there being some issue related to delta filter that made certain file formats [IIRC audio; maybe `.wav`] not extract properly on `SharpCompress`. So in my own  [Reloaded-II](https://github.com/Reloaded-Project/Reloaded-II); I had to swap that out for native `7z.dll` around 1.5 years ago. I kept looking and looking; 35 minutes later still empty handed 😅 .

Perf. wise I remember SharpCompress being considerably slower than native (maybe like 30%; I don't remember the exact figures). It might have improved since, I don't know. I do remember it specifically chugging *very hard* on archives with many tiny files though when SOLID compression was involved.

Anyway, as for other options...

# Native Library

Not really valid option, while `7z.dll` itself does have COM interface on Windows, the official Linux/Mac version(s) (`7zz`) don't provide an API. The unofficial port `p7zip` can be used as library IIRC, but that hasn't been maintained in at least 5 years.

I originally suggested this but @halgari mentioned a strange case where he got Access Violation (segfaults) using 7z.dll that never repro'd on 7z.exe; while I never ran into that; but I'll take their word on it.

# Native Binary

The issues listed in this issue.
Only thing I'd like to add; is starting new processes is slow and our current approach of using a binary makes extraction inefficient for cases where there are a lot of small mods.

-------

> This issue also is related to https://github.com/Nexus-Mods/NexusMods.App/pull/241

Maybe the best option is to just extract and re-compress in our own format in any case; there'll be cases where we will want to access more random data and extracting a SOLID block will take a while.

**Comment by halgari** [comment]
Right, just in general we have to trust that 7zip will *never* segfault, otherwise it'll crash the entire app process. Doing some sort of app containerization is another cross-platform nightmare.

As far as starting processes goes...yes it's slow, but it's also not (especially on *nix systems). Traditionally you'll never see more than 2000 mods in a single loadout, but the extraction overhead is much more dominated by disk IO and doing the actual work. 

That all being said, I'm very much in favor of a tiered extraction system. Use extractors (like ZipArchive) to support random access, use SharpCompress in the normal case, and fall back to 7zip when all else fails. This is the whole point behind the `Priority` return value on the `IExtractor`. 7zip can be `lowest` and everything else can override it.

**Comment by Aragas** [comment]
The tiered system will also allow to use better (faster) libraries for specific archieve types, when available. Plus do any other fine tuning to improve the performance of the most common formats
